src/main/java/org/opensearch/knn/index/codec/nativeindex/NativeIndexWriter.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.nativeindex;
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.codecs.DocValuesProducer;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.FieldInfo;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.index.VectorDataType;
+import org.opensearch.knn.index.codec.transfer.VectorTransfer;
+import org.opensearch.knn.index.codec.transfer.VectorTransferByte;
+import org.opensearch.knn.index.codec.transfer.VectorTransferFloat;
+import org.opensearch.knn.index.codec.util.KNNCodecUtil;
+import org.opensearch.knn.index.codec.util.SerializationMode;
+import org.opensearch.knn.index.util.KNNEngine;
+import org.opensearch.knn.indices.ModelCache;
+import org.opensearch.knn.plugin.stats.KNNGraphValue;
+
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.Value;
+
+import static org.opensearch.knn.common.KNNConstants.MODEL_ID;
+
+/**
+ * Abstract class to build the KNN index and write it to disk
+ */
+public abstract class NativeIndexWriter {
+
+    /**
+     * Class that holds info about vectors
+     */
+    @Builder
+    @Value
+    protected static class NativeVectorInfo {
+        private VectorDataType vectorDataType;
+        private int dimension;
+        private SerializationMode serializationMode;
+    }
+
+    /**
+     * Class that holds info about the native index
+     */
+    @Builder
+    @Value
+    protected static class NativeIndexInfo {
+        private FieldInfo fieldInfo;
+        private KNNEngine knnEngine;
+        private int numDocs;
+        private long arraySize;
+        private Map<String, Object> parameters;
+        private NativeVectorInfo vectorInfo;
+        private String indexPath;
+    }
+
+    protected final Logger logger = LogManager.getLogger(NativeIndexWriter.class);
+
+    /**
+     * Method for creating a KNN index in the specified native library
+     *
+     * @param fieldInfo
+     * @param valuesProducer
+     * @param indexPath
+     * @param isMerge
+     * @param isRefresh
+     * @throws IOException
+     */
+    public void createKNNIndex(FieldInfo fieldInfo, DocValuesProducer valuesProducer, String indexPath, boolean isMerge, boolean isRefresh)
```

Comment:
This is favoring polymorphism over composition. Not ideal since its used for code reuse

Reply:
I can change the classes to follow a composition-based design, but I think this is the way that @navneet1v wants it.

Reply:
Should we meet to discuss which design would be better?

Reply:
Its a basic principle which is often overlooked. Its a no brainer that composition should take precedence unless you are enforcing behavior. I don't think there is a big discussion required here. I will leave it up to @navneet1v and you to decide. 

As a rule I only use polymorphism if it ties to [Liskov substitution](https://en.wikipedia.org/wiki/Liskov_substitution_principle). Exceptions will be only when there is some sort of a constraint

https://en.wikipedia.org/wiki/Composition_over_inheritance#:~:text=Composition%20over%20inheritance%20(or%20composite,from%20a%20base%20or%20parent

Reply:
Ok. I think polymorphism would be better here since we can hide the internal type behind the abstract class so the reader doesn't have to worry about which NativeIndexWriter they are using.

It seems like this is a case Liskov subsititution. If we state the property is having the function createKNNIndex that makes an index and writes it to disk, then all subclasses share this property. That should be the only property of the class, since it holds no fields and only has one public function.

Reply:
On taking a deeper look I can see there are both pros and cons (one of the cons of composition here is the createIndex and write Index functions will not be used outside the codec, hence putting them in place makes sense), which will require further discussion on LLD of this class. Hence I propose we should merge this change in feature branch and then work on improving the LLD on the java side. Reason I want this change to be merged in feature branch is because this change is already became quite big. So lets just iterate over it in a feature branch. @shatejas please let me know your thoughts.

Reply:
@shatejas I have changed the branch to https://github.com/opensearch-project/k-NN/tree/feature/iterative-index-build for merging this feature.

Reply:
>  So lets just iterate over it in a feature branch

I don't mind

>  benefit in terms of performance

Best perfomance can be achieved by the most unmaintainable code. Its not about performance. We should sync up offline as this PR is not the right place

Reply:
Going to resolve this for now.

Reply:
> Best perfomance can be achieved by the most unmaintainable code. Its not about performance. We should sync up offline as this PR is not the right place

My main motive to ask about performance was if there is perf benefits then its no brainer that we should go with composition for sure. But if there is no, then its open for discussion. 

See one of the reason why I was on fence for composition here is the write and create index functionality is not needed outside of this writer as of now. Hence, putting them in 1 place provide a tighter control. But if there are use-cases of those composition else where then I agree we should go with composition. Hence lets have a gh issue for this discussion.

Reply:
> if there is perf benefits then its no brainer

Maintainability of the code is also a no brainer, especially when its not affecting performance

> I was on fence for composition here is the write and create index functionality is not needed outside of this writer as of now.

There are ways to do it. private-package with strategy pattern is one. The problem here is has a potential to violate [OCP](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle). Its open to modification, an implementation can simply override the public createIndex since the class cannot be made final. Composition generally helps to stick to [SRP](https://en.wikipedia.org/wiki/Single-responsibility_principle) and makes testing easier. For instance, it forces to test public createIndex in unit tests for every implementation, composition simplifies that.